### ISSUE TITLE
node-inspector and babel-register

### ISSUE THREAD

**Issue by mdotasia** [node-inspector/node-inspector#832]
I am using this setup https://babeljs.io/docs/setup/#mocha to run mocha tests written in ES6. I would like to debug a mocha test with node-inspector. Running ES6 mocha tests works fine with this setup.

In order to debug, I run `node-inspector` in one terminal and

```
mocha --compilers js:babel-register --debug-brk
```
 in another. I can debug in principle, i.e. on any ES5 code I can set break points. Any breakpoints in my ES6 tests get ignored. It looks like node-inspector does not pick up any sourcemaps.

Now babel-register does create sourcemaps by default, only I have no idea whether or not it saves compiles files to disk and whether node-inspector has a chance to get its hand on the sourcemaps.

Has anyone managed to debug code in node-inspector that has been processed with babel-register?

**Comment by mdotasia** [comment]
I did some more analysis. It turns out that on Windows 7 with Node 5.6.0 I can debug mocha tests without problem. When I debug a `test.js` (ES6) I launch node-inspector/mocha as described in my previous comment in Chrome. Then I can set breakpoints in `test.js`. I press continue and then I see that node-inspector replaces `test.js` with the compiled versions and a `test.js.source` appears, which contains the original source ES6 code. In `test.js.source` I can set additional breakpoints and all is good.

Now I switch to my Mac and do exactly the same thing. In Chrome the test exits with an error. In Opera I have more luck and I can launch the debug session and set a breakpoint in `test.js`. I press continue and the debugger stops in the compiled file. The original ES6 file does not seem to be linked to the compiled file, even though I see inline sourcemaps. None of the magic that happened in Windows seems to happen on my Mac.

What's going wrong here?

**Comment by 3y3** [comment]
Hm... Last time then I fix source maps I make if for Win.
Maybe now I need to inspect problems for Mac.
I don't promise that this will be fast - in current time I intentionally work on major release.

Is your issue happens only for main file or also for all sub scripts?

**Comment by mdotasia** [comment]
In above scenario I'm running my tests with

```
mocha --recursive --compilers js:babel-register
```

and then I'm trying to debug only the test that failed. Weirdly on Windows it work all fine. But Mac seems to be broken.

**Comment by mdotasia** [comment]
I played around with `node-inspector` debugging a bit more and tested debugging with normal babel compilation, i.e. not using `babel-register` for on-the-fly compilation but **compile first** and **then debug**. Here is what I found for debugging with `node-inspector`:

- Windows: debugging with
  - `babel-register` (implicit inline source maps): works
  - `babel -s` (source maps in separate files): works
  - `babel -s inline` (inline source maps): does not work
- Mac: debugging with
  - `babel-register` (implicit inline source maps): does not work
  - `babel -s` (source maps in separate files): works
  - `babel -s inline` (inline source maps): works

**Comment by DiknT** [comment]
same issue here with babel-register on OS X  while it is working fine on windows

**Comment by itaylor** [comment]
I'm running mocha tests using babel-register on Mac and am seeing the same issues.   I don't hit my breakpoints in node-inspector for any files that are have been transpiled by babel.  Node 5.8.0, babel-register 6.7.2.

**Comment by AndrejGajdos** [comment]
+1 (babel-register on OS X)

**Comment by papaspiridis** [comment]
I can confirm this is happening too.

The strange thing is that on a clean install it worked fine initially but after a while started messing up.
I noticed the following behavior:

- Running node-inspector separately returns the transpiled code
- Running node-debug picks up the original files (but I can't get the program to execute)

**Comment by 3y3** [comment]
Should be fixed in 0.12.8

**Comment by flying-sheep** [comment]
doesn’t work for me.

i’m using `mocha.opts` to specify `--compilers js:./my-babel-register.js --recursive`

my-babel-register.js:

```js
require('babel-register')({
  ignore: /\/(dist|node_modules)\//
})
```

when i run `node_modules/.bin/node-debug node_modules/.bin/_mocha`, i end up in “modules.js”, on the line

```js
var result = compiledWrapper.apply(this.exports, args);
```

break points don’t work, i can only run the tests to the end.

**Comment by alexanderkartavtsev** [comment]
+1
the same issue with v.0.12.8 on OS X

**Comment by wafisher** [comment]
@mdotasia Do you mind sharing exactly how you compile your ES6 code and how you test it? I can't seem to get the combination of stuff right to get Babel + Mocha + node-inspector working in order to debug ES6 code with Mocha. Thanks!

**Comment by bochen2014** [comment]
@wafisher 
@mdotasia mentioned a couple of ways of compiling es6 using babel and then testing it with mocha.

the easist way (and most common way) is either
1. mocha --require setup.js  your_test.js  , while  in setup.js, you have  require('babel-register')(); 
2. mocha --compilers js:babel-register your_test.js 

you open the first terminal running node-inspector, then you startup mocha with --debug-brk, and it works on windows . I've tested it today, works fine (there is some performance issue, which I raised a issue #929 

after I read @mdotasia 's description, I tried the second approach, which is to pre-compile/transpile es6 to es5 using babel cli. 
```
$ babel ./ -d compiled --source-map --ignore node_modules 
```
which will transpile your es6 source code and put it into compiled dir.

you then follow the same process to debug it. and performance is improved. I have also verified this.


I just have one question: why node-debug doesn't work??? why must run node-inspector in separate teminal??
